### DISCUSSION TITLE
dynamic select without knowing the table name

### DISCUSSION THREAD

**Discussion by chekaiyin** [uptrace/bun#691] [Q&A] [answered]
I need a universal function to retrieve all records from tables.
Struct:
```
type Account struct {
	bun.BaseModel

	ID    int64  `bun:",pk"`
	Name  string `bun:",notnull"`
	Owner string `bun:",notnull"`
}

func (*Account) Codes() []interface{} {
	return []interface{}{
		Account{ID: 1, Name: "name1", Owner: "owner1"},
		Account{ID: 2, Name: "name2", Owner: "owner2"},
	}
}

type CodeTable interface {
	Codes() []interface{}
}
```

I would like to call the function this way:
`loadCodeTable(db, (*Account)(nil))`

in function loadCodeTable, it needs to retrieve all records and returning to a slice
it works with map:
```
var dbCodes []map[string]interface{}
err := db.NewSelect().Model(codeTable).Scan(context.Background(), &dbCodes)
```

but how can I retrieve all records to a slice of struct Account? I tried reflect, but no help.

Thanks!

**Comment by jhhom** [comment, marked as answer]
Hi @chekaiyin, if I understand correctly, you need a universal `loadCodeTable` that can be called on slices of any structs, not just `Account` struct.

The following is what I came up with, not sure if it solves your problem.

You can retrieve all records to the `Account` struct by simply passing a reference of the slice to the `Scan` function like so:
```go
func loadCodeTable2(db *bun.DB, codeTable CodeTable) ([]Account, error) {
	accounts := make([]Account, 0)
	err := db.NewSelect().Model(codeTable).Scan(context.Background(), &accounts)
	return accounts, err
}
```


If you need the function to be universal across every type, not just for `Account`, you can use  Golang generics.
```go
func loadCodeTable[T any](db *bun.DB, codeTable CodeTable) ([]T, error) {
	data := make([]T, 0)
	err := db.NewSelect().Model(codeTable).Scan(context.Background(), &data)
	return data, err
}
```


You can then use the above generic function for different types of struct, not just for `Account`, to do so you just need to pass the type parameter when you call it:
```go
data, err := loadCodeTable[Account](db, (*Account)(nil))

data2, err := loadCodeTable[Person](db, (*Person)(nil))
```

**Reply by chekaiyin** [reply]
thanks @jhhom it resolved the issue!
